### PR TITLE
Warning period for detecting nested impl trait

### LINKED ISSUE #57979: Regression from stable to nightly: nested impl trait is not allowed
I have some code that looks a bit like this:

```rust
use std::borrow::Borrow;

pub struct Data<TBody>(TBody);

pub fn collect(_: impl IntoIterator<Item = impl Borrow<Data<impl AsRef<[u8]>>>>) {
    unimplemented!()
}
```

On `1.32.0` (and the current beta `1.33.0-beta.4 2019-01-24 635817b9db20ecdcd036`) this compiles successfully.

On the current nightly `1.34.0-nightly 2019-01-28 d8a0dd7ae88023bd09fa` this fails with:

```
error[E0666]: nested `impl Trait` is not allowed
 --> src/main.rs:5:61
  |
5 | pub fn collect(_: impl IntoIterator<Item = impl Borrow<Data<impl AsRef<[u8]>>>>) {
  |                                            -----------------^^^^^^^^^^^^^^^^--
  |                                            |                |
  |                                            |                nested `impl Trait` here
  |                                            outer `impl Trait`
```

A workaround is just to use generics. Did we mean to change the rules around nested impl trait here?

### PR DESCRIPTION
**Pull request by pnkfelix**
Here is some proposed code for making a warning period for the new checking of nested impl trait.

It undoes some of the corrective effects of PR #57730, by using boolean flags to track parts of the analysis that were previously skipped prior to PRs #57730 and #57981 landing.

Cc #57979

### CODE REVIEW

**Comment by pnkfelix** [comment]
nominating for discussion at T-compiler meeting, in terms of evaluating the cost/benefit of doing a warning cycle in this case.

**Comment by Centril** [comment]
I think the benefit of a warning cycle is low (don't think this is too common) but I think so is the cost since we don't want to use this syntax for anything *right now*. That said, I'm wary of the number of [C-future-compatibility](https://github.com/rust-lang/rust/labels/C-future-compatibility) lints that we have right now and that some of them have existed for so long.

To make sure these don't spiral out of control and to bring them down to a more manageable number, I'd like us to set stricter timelines and deadlines for turning these warnings into deny-lints and then to hard errors. The timelines can be set on a case-by-case basis when we make a future-compat lint. E.g. if we merge this before 1.34 goes into beta, in this instance I think it should become `#[deny(..)]` by 1.35 (+? 1 release) and a hard error by 1.36 (+? 1 release).

**Comment by pnkfelix** [comment]
@Centril note that the PR as written does not make use of the lint infrastructure.

I suppose I should interpret your comment as pushing for the PR to be changed to be a proper lint, yes?

**Comment by Centril** [comment]
@pnkfelix I assumed this was just a quick hack and that you'd convert it to the infrastructure we have for C-future-compatibility warnings (and with a dedicated issue). You should interpret my comment as "we should schedule times for when we progress C-future-compatibility to deny to errors" starting with this one.

**Comment by pnkfelix** [comment]
Okay. I'll wait until we discuss this at T-compiler meeting; if we collectively decide that we will do a warning-cycle, then I'll implement a proper lint. If we decide that we won't, then I'll just close this PR then.

**Comment by Zoxc** [comment]
This is changes existing errors into warnings too though, instead of just the edge case which was allowed.

**Comment by nikomatsakis** [comment]
My feeling:

- Landing a lint is probably the *right* thing to do
- But if it is just a single (or a few) affected crates, and those crates have moved on, I'm not sure it's worth it
- I agree that we should do a better job reviewing future compatibility warnings and generally establishing fixed timelines for them etc, I think that's a good topic to review as part of the Meta WG

**Comment by nikomatsakis** [comment]
Oh, if we were going to do it, I would prefer to do it as a "real future compat lint"

**Comment by pnkfelix** [comment]
@Zoxc wrote:

> This is changes existing errors into warnings too though, instead of just the edge case which was allowed.

Hmm, my intention was to only catch the cases where we went injected fresh errors. I may have messed that up (as in, I may have misread the original code prior to PR #57730 and thought it was missing more than it was in its recursive traversal).

I'll try to double-check the situation there. I certainly don't want to downgrade more errors to warnings than necessary.

File: src/librustc_passes/ast_validation.rs [hunk 1]
```diff
@@ -36,6 +36,11 @@ struct AstValidator<'a> {
     // Used to ban `impl Trait` in path projections like `<impl Iterator>::Item`
     // or `Foo::Bar<impl Trait>`
     is_impl_trait_banned: bool,
+
+    // rust-lang/rust#57979: the ban of nested `impl Trait` was incomplete
+    // until sometime after PR #57730 landed. We track whether we should
```
Comment (Zoxc) [inline src/librustc_passes/ast_validation.rs]: I think should say `bugged until PR #57730 landed`.

**Comment by Zoxc** [comment]
I'd suggest bringing back `NestedImplTraitVisitor` from before https://github.com/rust-lang/rust/pull/57730 and making a list of nodes which should have errors instead of warnings from that. Then pass that list on to `AstValidator`.

**Comment by pnkfelix** [comment]
> I'd suggest bringing back `NestedImplTraitVisitor` from before #57730 and making a list of nodes which should have errors instead of warnings from that. Then pass that list on to `AstValidator`.

yes okay I think I see where I went wrong now. I was doing this too late at night.

**Comment by pnkfelix** [comment]
I wrote a long comment on [one of the commits](https://github.com/rust-lang/rust/pull/58608/commits/d61d14bfcb64510c62a2ba8f94854702808ca3b7), but this is the important bit:

this version is an "improvement" on the one I posted earlier, so I am going to share it
the following reasons:

 1. Maybe someone will point out where I have gone wrong, either in my    analysis above or in the code itself

 2. This version is clearly more complex than the (bad) one I had    proposed earlier. And it may get even more complex before its    "right." This all influences the debate on whether we attempt to    land a warning cycle here at all.

**Comment by pnkfelix** [comment]
also: It is of course a red flag that I have added two different pieces of boolean state for tracking warning cycle state, but only *one* diagnostic has changed in our test output.

My current hypothesis is that this is a reflection of weakness in our test suite, rather than a redundancy in the approach. But it would be good to actually come up with a concrete test ilustrating the point.

**Comment by pnkfelix** [comment]
> My current hypothesis is that this is a reflection of weakness in our test suite, rather than a redundancy in the approach. But it would be good to actually come up with a concrete test illustrating the point.

Here is a concrete example illustrating the case I am trying to catch with the second boolean flag:

```rust
pub trait Bar { }

pub trait Quux<T> { type Assoc; }
pub fn demo(_: impl Quux<(), Assoc=<() as Quux<impl Bar>>::Assoc>) { }

impl<T> Quux<T> for () { type Assoc = u32; }

fn main() { }
```

**Comment by pnkfelix** [comment]
Okay, I think I'm done working on this.

However I believe I have also missed the cut-off for beta-next (see #58747).
 * (Clearly I should have devoted more time to this yesterday. Or dropped it entirely...)

Anyway, assuming that we are not planning to recut beta, there isn't much point in landing this *unless* we also plan to evenutally beta-backport it.

So, nominating for discussion (again) at T-compiler meeting.

**Comment by pnkfelix** [comment]
(I'm beta-nominating mostly as a procedural matter. I personally am not convinced that a beta-backport is warranted here.

**But**: if we do not intend to beta-backport, then this PR probably should not land at all, since I do not think there is much value in having stable-breakage followed by a 6-week warning cycle. Or at least, we could just just as easily just let the trains run, and if there *is* significant stable breakage reported, then we do both beta- and stable-back ports at that time, and have a point release.

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. We are pre-disposed to beta-backport, but we want to wait a week because we want to give people a chance to object asynchronously.

**Comment by pnkfelix** [comment]
heh I added a method but its only called from `debug!`, so our dead_code lint flagged it. Will fix.

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. beta-accepting, but please do not initiate backport until review has actually happened.

**Comment by pnkfelix** [comment]
during T-compiler meeting discussion @Centril [pointed out](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/weekly.20meeting.202019-03-07.20.2354818/near/160206215) that I need to file an associated C-future-compatibility issue for this. I don't expect that addition to change the diff here all that much, but I figure I should at least mention it and mark this as WIP until that is done.

**Comment by Zoxc** [comment]
This looks good to me. One thing to note is that the `warning_period_57979_nested_impl_trait` field is sticky, which means that things that are errors on stable could become warning here, before being turned back into errors again in some future release.

**Comment by Centril** [comment]
(We should not turn errors on stable into warnings...)

**Comment by pnkfelix** [comment]
I did in an earlier draft of this PR try to avoid making the booleans sticky; in particular, I had tried to reset them back to `false` at what I thought was the point in the control-flow that corresponded to where the old code would have actually re-initiated the recursive checking.

However, that attempt at a reset back to `false` didn't seem to be in the correct spot, so I decided that I must have misunderstood the control flow.

I'll look again and try to make the fields not-sticky.

**Comment by Zoxc** [comment]
@pnkfelix I think you'll need to bring back `NestedImplTraitVisitor` in order to do that. `warning_period_57979_impl_trait_in_proj` should be sticky, so that's fine.

**Comment by pnkfelix** [comment]
I just want to note, we have a set of potentially conflicting factors here:

> (We should not turn errors on stable into warnings...)

Similarly, we should not inject hard-errors into stable without a warning cycle (which is what motivated this PR in the first place).

And we should not overly complicate our own code base (which is why I've been trying to avoid solutions like bringing back `NestedImplTraitVisitor` itself whole hog). Note that I'm very much on the fence about how much these boolean flags as written are complicating the code base as it stands.

----

I don't know how to weigh the relative risk of the two options I see here; that is, the risk of:
  1. injecting hard errors into stable versus
  2. turning stable-errors into warnings.

*Neither* of the two options is great.

But also, neither option is ... that bad (?), given that we don't expect the given coding pattern to be very common in the first place. Plus I think the outcome that result from a downgrade from stable-error-to-warning are not unsoundness, but rather either fragile-dependence on type-inference details, or ICE'ing?

----

I haven't come to any concrete conclusions yet. I'm going to spend a little bit of time investigating (again) the option of making `warning_period_57979_nested_impl_trait` non-sticky. If I again fail to come up with something satisfactory, then I'll have to decide whether to move ahead with this PR or to close it.

**Comment by pnkfelix** [comment]
@Zoxc you mind taking a look at this revised (non-sticky) approach for the nested `impl Trait` check?

**Comment by Zoxc** [comment]
This looks good to me now, though there's still the incorrect comment I noted earlier.

**Comment by pnkfelix** [comment]
> This looks good to me now, though there's still the incorrect comment I noted earlier.

Discussed in [zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/nested.20impl.20trait.20warning.20period.20PR.20.2358608/near/160563406); I will address.

**Pull request merged**